zkevm-circuits/src/evm_circuit.rs

```diff
@@ -506,3 +506,82 @@ pub mod test {
         run_test_circuit(block, FixedTableTag::iter().collect())
     }
 }
+
+#[cfg(test)]
+mod evm_circuit_stats {
+    use super::test::*;
+    use super::*;
+    use crate::evm_circuit::step::ExecutionState;
+    use eth_types::{bytecode, evm_types::OpcodeId, geth_types::GethData};
+    use halo2_proofs::pairing::bn256::Fr;
+    use halo2_proofs::plonk::ConstraintSystem;
+    use mock::test_ctx::{helpers::*, TestContext};
+    use strum::IntoEnumIterator;
+
+    /// This function prints to stdout a table with all the implemented states
+    /// and their responsible opcodes with the following stats:
+    /// - height: number of rows used by the execution state
+    /// - gas: gas value used for the opcode execution
+    /// - height/gas: ratio between circuit cost and gas cost
```

Comment:
not sure why need this ratio ?  using  each step(op code)'s gas / total gas makes more sense ? then we can know each step gas cost ratio clearly.

Reply:
The reason to include this ratio in the table is to figure out which opcodes present the worst case when limiting the gas block.

In general we will have a circuit setup with a given degree, and we'll need to specify a maximum gas per block for network.  The maximum gas per block is determined by the execution trace that manages to use the maximum number of rows with the minimum amount of gas.  If we set the maximum gas per block higher than that, we could have a user send a tx for which we can't build a proof because it needs more rows than `2^k`.  So having the ratio between height / gas is useful because it allows us to see which circuit opcode to optimize to increase the maximum gas per block (we take the opcode with higher ratio and try to reduce the number of rows). 

The exercise in this PR only applies to the EVM circuit, but we can do similar ones to the other circuits.  For example, which is the opcode that manages to use the most rw_table rows with the least amount of gas?  See also this discussion about this kind of analysis https://github.com/privacy-scaling-explorations/zkevm-specs/issues/71

Reply:
understood !!